### PR TITLE
Ka 185

### PR DESCRIPTION
**Pull request by FredPeal**

### CODE REVIEW

File: app/Console/Commands/UpdateAbilitiesCommand.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace App\Console\Commands;
+
+use Illuminate\Console\Command;
+use Kanvas\AccessControlList\Actions\CreateAbilitiesByModule;
+use Kanvas\Apps\Models\Apps;
+
+class UpdateAbilitiesCommand extends Command
```
Comment (kaioken) [inline app/Console/Commands/UpdateAbilitiesCommand.php:9]: @FredPeal move this to a folder under ACL or AccessControl

File: app/GraphQL/Ecosystem/Mutations/Roles/RolesManagementMutation.php [hunk 2]
```diff
@@ -69,6 +69,14 @@ public function removeRoleFromUser(mixed $rootValue, array $request): bool
         return true;
     }
 
+    public function givePermissionToRole(mixed $rootValue, array $request): bool
+    {
+        $systemModule = SystemModulesRepository::getByModelName($request['systemModule']);
```
Comment (kaioken) [inline app/GraphQL/Ecosystem/Mutations/Roles/RolesManagementMutation.php:74]: @FredPeal pass the curernt app as param

File: app/GraphQL/Ecosystem/Mutations/Roles/RolesManagementMutation.php [hunk 3]
```diff
@@ -85,7 +93,7 @@ public function givePermissionToUser(mixed $rootValue, array $request): bool
             $user = UsersRepository::getUserOfCompanyById($company, $userId);
         }
 
-        Bouncer::allow($user)->to(Str::slug($request['permission']));
+        Bouncer::allow($user)->to($request['permission']);
```
Comment (kaioken) [inline app/GraphQL/Ecosystem/Mutations/Roles/RolesManagementMutation.php:96]: @FredPeal shouldn't we always want permission to follow a slug pattern ?

File: database/migrations/2024_04_15_162222_abilities_modules_table.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+return new class () extends Migration {
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        Schema::create('abilities_modules', function (Blueprint $table) {
+            $table->id();
+            $table->integer('system_modules_id');
```
Comment (kaioken) [inline database/migrations/2024_04_15_162222_abilities_modules_table.php:15]: @FredPeal add apps_id

File: app/GraphQL/Ecosystem/Queries/Roles/RoleAbilitiesQuery.php [hunk 5]
```diff
@@ -21,4 +24,22 @@ public function getAllAbilities(mixed $root, array $query): array
 
         return $mapAbilities->all();
     }
+
+    public function getAllAbilitiesByRoles(mixed $root, array $request)
+    {
+        $roles = Bouncer::role()->where('name', $request['role'])->first();
```
Comment (kaioken) [inline app/GraphQL/Ecosystem/Queries/Roles/RoleAbilitiesQuery.php:30]: @FredPeal make sure to filter this by app

File: src/Kanvas/AccessControlList/Enums/ModuleEnum.php [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Kanvas\AccessControlList\Enums;
+
+enum ModuleEnum: int
```
Comment (kaioken) [inline src/Kanvas/AccessControlList/Enums/ModuleEnum.php:7]: @FredPeal this should be global not in acl

**Review by kaioken — CHANGES_REQUESTED** [review]

File: app/GraphQL/Ecosystem/Queries/Roles/RoleAbilitiesQuery.php (hunk 5 above)
Comment (FredPeal) [reply app/GraphQL/Ecosystem/Queries/Roles/RoleAbilitiesQuery.php:30]: Bouncer apply the filter on background

File: app/GraphQL/Ecosystem/Queries/Roles/RoleAbilitiesQuery.php [hunk 7]
```diff
@@ -21,4 +24,22 @@ public function getAllAbilities(mixed $root, array $query): array
 
         return $mapAbilities->all();
     }
+
+    public function getAllAbilitiesByRoles(mixed $root, array $request)
```
Comment (kaioken) [inline app/GraphQL/Ecosystem/Queries/Roles/RoleAbilitiesQuery.php:28]: @FredPeal return type

**Review by kaioken — CHANGES_REQUESTED** [review]

File: src/Kanvas/AccessControlList/Actions/CreateAbilitiesByModule.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+use Kanvas\AccessControlList\Models\AbilitiesModules;
+use Kanvas\AccessControlList\Repositories\ModulesRepositories;
+use Kanvas\Apps\Models\Apps;
+use Kanvas\SystemModules\Repositories\SystemModulesRepository;
+
+class CreateAbilitiesByModule
+{
+    protected Apps $app;
+
+    public function __construct(?Apps $app = null)
```
Comment (kaioken) [inline src/Kanvas/AccessControlList/Actions/CreateAbilitiesByModule.php:19]: @FredPeal you can do procted ?apps $app = null in the constructor

File: src/Kanvas/AccessControlList/Repositories/ModulesRepositories.php [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+use Kanvas\Companies\Models\Companies;
+use Kanvas\Users\Models\Users;
+use Kanvas\Inventory\Products\Models\Products;
+use Kanvas\Inventory\Products\Models\ProductsTypes;
+use Kanvas\Inventory\Regions\Models\Regions;
+use Kanvas\Inventory\Warehouses\Models\Warehouses;
+
+class ModulesRepositories
+{
+    public static function getAbilitiesByModule(): array
```
Comment (kaioken) [inline src/Kanvas/AccessControlList/Repositories/ModulesRepositories.php:18]: @FredPeal thi s more a template than a repository , or enum, I wouldn't lvea this class in the repository folder

File: src/Kanvas/AccessControlList/Repositories/RolesRepository.php [hunk 10]
```diff
@@ -59,4 +65,60 @@ public static function getAppRoles(AppInterface $app): Collection
     {
         return Role::where('scope', RolesEnums::getScope($app))->get();
     }
+
+    public static function getMapAbilityInModules(string $roleName): array
+    {
+        $roles = Bouncer::role()->where('name', $roleName)->firstOrFail();
```
Comment (kaioken) [inline src/Kanvas/AccessControlList/Repositories/RolesRepository.php:71]: @FredPeal dont you need to filter the role by app?

File: app/GraphQL/Ecosystem/Mutations/Roles/RolesManagementMutation.php [hunk 11]
```diff
@@ -70,6 +72,28 @@ public function removeRoleFromUser(mixed $rootValue, array $request): bool
...
+        return true;
+    }
+
+    public function removePermissionToRole(mixed $rootValue, array $request): bool
+    {
+        $systemModule = SystemModulesRepository::getByModelName($request['systemModule'], app(Apps::class));
+        Bouncer::disallow($request['role'])->to($request['permission'], $systemModule->model_name);
+
+        $roles = RolesRepository::getMapAbilityInModules($request['role']);
+        Redis::set('roles:abilities', $roles);
```
Comment (kaioken) [inline app/GraphQL/Ecosystem/Mutations/Roles/RolesManagementMutation.php:92]: @FredPeal i would move the roles:abilities to a enum so I don't have to worry about that string in 2 diff places

File: src/Kanvas/Enums/ModuleEnum.php [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+declare(strict_types=1);
+
+namespace Kanvas\Enums;
+
+enum ModuleEnum: int
+{
+    case ECOSYSTEM = 1;
+    case INVENTORY = 2;
+    case CRM = 3;
```
Comment (kaioken) [inline src/Kanvas/Enums/ModuleEnum.php:11]: add social, workflow, action engine

**Review by kaioken — CHANGES_REQUESTED** [review]

File: src/Kanvas/AccessControlList/Repositories/RolesRepository.php (hunk 10 above)
Comment (FredPeal) [reply src/Kanvas/AccessControlList/Repositories/RolesRepository.php:71]: No Bouncer Apply the scope into the package

**Pull request closed without merging**